pandas/plotting/_matplotlib/core.py

```diff
@@ -120,7 +121,77 @@ def __init__(
 
         self.sort_columns = sort_columns
 
-        self.subplots = subplots
+        if isinstance(subplots, bool):
+            self.subplots = (
+                subplots
+            )  # type: Optional[Union[bool, Sequence[Sequence[Union[str, int]]]]]
+        elif isinstance(subplots, Iterable):
+            supported_kinds = (
+                "line",
+                "bar",
+                "barh",
+                "hist",
+                "kde",
+                "density",
+                "area",
+                "pie",
+            )
+            if self._kind not in supported_kinds:
+                raise ValueError(
+                    "When subplots is an iterable, kind must be "
+                    f"one of {', '.join(supported_kinds)}. Got {self._kind}."
+                )
+
+            if any(
+                not isinstance(group, Iterable) or isinstance(group, str)
+                for group in subplots
+            ):
+                raise ValueError(
+                    "When subplots is an iterable, each entry "
+                    "should be a list/tuple of column names."
+                )
+
+            cols_in_groups = [col for group in subplots for col in group]
+            duplicates = {
+                col for (col, cnt) in Counter(cols_in_groups).items() if cnt > 1
+            }
+            if duplicates:
+                raise ValueError(
+                    f"Each column should be in only one subplot. Columns {duplicates} "
+                    "were found in multiple sublots."
+                )
+
+            cols_in_groups_set = set(cols_in_groups)
```

Comment:
I would rather just do `cols_in_groups = set(cols_in_groups)` but I couldn't make pypy happy otherwise

Reply:
Again, if you want to reuse variable names for different types, needs a variable annotation before first assignment. I think OK to use different name.